pgarchive

```diff
@@ -0,0 +1,1039 @@
+                    This is sourced as a shell script, and could theoretically be used to override
+                    most internal variables and functions, at your own risk.
+
+EOF
+}
+
+bash_completion() {
+    cat <<'EOF'
+# pgarchive autocompletion for bash
+_pgarchive() {
+    local cur cmd subcmd cmds context
+    COMPREPLY=()
+    cur="${COMP_WORDS[COMP_CWORD]}"
+    cmd=""
+    subcmd=""
+    [[ $COMP_CWORD -ge 2 ]] && cmd="${COMP_WORDS[1]}"
+    [[ $COMP_CWORD -ge 3 ]] && subcmd="${COMP_WORDS[2]}"
+
+    declare -A cmds
+    cmds["/"]="help container wal_archive standby snapshot clone"
+    cmds["container/"]="start stop status dashboard init init-cron purge"
+    cmds["wal_archive/"]="start stop status list log"
+    cmds["standby/"]="start stop status restart reload log"
+    cmds["snapshot/"]="create delete expire list"
+    cmds["snapshot/create"]="--force --no-force"
+    cmds["snapshot/delete"]="SNAPSHOT"
+    cmds["snapshot/expire"]="--show"
+    cmds["clone/"]="list create restore duplicate start stop status reload restart delete log psql"
+    cmds["clone/list"]="--status"
+    cmds["clone/create"]="SNAPSHOT"
+    cmds["clone/duplicate"]="CLONE"
+    cmds["clone/start"]="CLONE"
+    cmds["clone/stop"]="CLONE_OR_ALL"
+    cmds["clone/status"]="CLONE"
+    cmds["clone/reload"]="CLONE"
+    cmds["clone/restart"]="CLONE"
+    cmds["clone/delete"]="CLONE"
+    cmds["clone/log"]="CLONE"
+    cmds["clone/psql"]="CLONE"
+
+    context=${cmds["${cmd}"/"${subcmd}"]} || return 1
+    case "$context" in
+        SNAPSHOT)
+            COMPREPLY=($(compgen -W "`ls $PGARCHIVE/snapshots`" -- ${cur}))
+            ;;
+        CLONE)
+            COMPREPLY=($(compgen -W "`ls $PGARCHIVE/clones`" -- ${cur}))
+            ;;
+        CLONE_OR_ALL)
+            COMPREPLY=($(compgen -W "`ls $PGARCHIVE/clones` --all" -- ${cur}))
+            ;;
+        *)
+            COMPREPLY=($(compgen -W "${context}" -- ${cur}))
+            ;;
+    esac
+    return 0
+}
+complete -F _pgarchive pgarchive
+EOF
+}
+
+# For internal use only, by standby/recovery.conf.
+# This restore_command script can handle plain and gzipped files.
+# This is executed without the usual setup() ceremony, by postgres, with cwd=$PGDATA.
+standby_restore() {
+
+    wal="../wal_archive"
+    src="$wal/$1"
+    zsrc="${src}.gz"
+    dst="$2"
+
+    [[ -d "$wal" ]] || fail "The wal_archive directory does not exist, something is very wrong."
+
+    cp --reflink=always "$src" "$dst" 2>/dev/null || zcat "$zsrc" > "$dst" 2>/dev/null
+}
+
+# For internal use only, by standby/recovery.conf.
+# This is executed without the usual setup() ceremony, by postgres, with cwd=$PGDATA.
+# Note the $cleanup_position file is used by the wal_archive cronjob zipper, snapshotting,
+# and expiry.
+standby_record_cleanup_position() {
+
+    cleanup_position="pgarchive_cleanup_position"
+    restartpoint="$1"
+
+    # Note: This rm works around a quirk in our monitoring setup.
```

Comment:
Yes, so none removes it, since it's not strictly necessary because of the echo below.